### PR TITLE
[FIX] odoo-print: Don't print if zero copies are requested.

### PR DESCRIPTION
**Pull request by kdwyer**
story/13810

Signed-off-by: Kevin Dwyer <kevin.dwyer@unipart.io>

### CODE REVIEW

File: addons/print/tests/test_print_printer.py [hunk 1]
```diff
@@ -408,3 +409,18 @@ def test27_user_one_default_group_per_report_type(self):
...
+        # directly.
+        with self.assertLogs('odoo.addons.print.models.print_printer', level=logging.INFO) as cm:
+            self.printer_default.spool_report(self.printer_default.ids,
+                                              'print.report_test_page',
+                                              copies=0)
+            self.mock_subprocess.Popen.assert_not_called()
+            self.mock_subprocess.Popen.reset_mock()
+        self.assertEqual(
+            cm.output,
+            ['INFO:odoo.addons.print.models.print_printer:Zero copies requested, nothing will be printed.'])
```
Comment (armandcela) [inline addons/print/tests/test_print_printer.py]:
```suggestion
            ["INFO:odoo.addons.print.models.print_printer:Zero copies requested, nothing will be printed."])
```

File: addons/print/tests/test_print_printer.py [hunk 2]
```diff
@@ -408,3 +409,18 @@ def test27_user_one_default_group_per_report_type(self):
 
         # Default PDF printer should be unaffected
         self.assert_printer_is_user_default(self.group_upstairs, self.user_alice)
+
+    def test28_abandons_print_if_zero_copies_requested(self):
+        """If zero copies are requested not print should take place."""
+        # Odoo changes logging.INFO to 25 in netsvc.py; it seems the only way
+        # to assert the log level is to reference the module attribute
+        # directly.
+        with self.assertLogs('odoo.addons.print.models.print_printer', level=logging.INFO) as cm:
```
Comment (armandcela) [inline addons/print/tests/test_print_printer.py]:
```suggestion
        with self.assertLogs("odoo.addons.print.models.print_printer", level=logging.INFO) as cm:
```

File: addons/print/tests/test_print_printer.py [hunk 3]
```diff
@@ -408,3 +409,18 @@ def test27_user_one_default_group_per_report_type(self):
 
         # Default PDF printer should be unaffected
         self.assert_printer_is_user_default(self.group_upstairs, self.user_alice)
+
+    def test28_abandons_print_if_zero_copies_requested(self):
+        """If zero copies are requested not print should take place."""
+        # Odoo changes logging.INFO to 25 in netsvc.py; it seems the only way
+        # to assert the log level is to reference the module attribute
+        # directly.
```
Comment (armandcela) [inline addons/print/tests/test_print_printer.py]: I think the comment should be inside the docstring

File: addons/print/models/print_printer.py [hunk 4]
```diff
@@ -182,6 +182,10 @@ def spool_report(self, docids, report_name, data=None, title=None, copies=1):
         """Spool report to printer"""
         # pylint: disable=too-many-arguments, too-many-locals
 
+        if copies == 0:
+            _logger.info(_('Zero copies requested, nothing will be printed.'))
```
Comment (armandcela) [inline addons/print/models/print_printer.py]:
```suggestion
            _logger.info(_("Zero copies requested, nothing will be printed."))
```
Black Formatting

File: addons/print/models/print_printer.py [hunk 5]
```diff
@@ -182,6 +182,10 @@ def spool_report(self, docids, report_name, data=None, title=None, copies=1):
         """Spool report to printer"""
         # pylint: disable=too-many-arguments, too-many-locals
 
+        if copies == 0:
```
Comment (PeterC10) [inline addons/print/models/print_printer.py]: Suggestion: Might be worth changing this to `if not copies:`. If `copies` is set to `None` then the print wouldn't be abandoned.

Although can't think why anyone would set `copies` to `None` anyway so feel free to resolve if you don't think it's necessary. :)

File: addons/print/tests/test_print_printer.py (hunk 1 above)
Comment (kdwyer) [reply addons/print/tests/test_print_printer.py]: In general this addon uses single quotes for strings, so it seems best to keep it consistent for now?

Also, it's a fork, so blackening would be a significant change from the base repo.

**Review by armandcela — APPROVED** [review]

**Comment by armandcela** [comment]
Resolved even the suggestion for having only one docstring and removing the comment. Do what you feel is better about it.

**Comment by kdwyer** [comment]
> Resolved even the suggestion for having only one docstring and removing the comment. Do what you feel is better about it.

The comment is about an implementation detail of the test code, so I don't think it belongs in the docstring.

File: addons/print/models/print_printer.py (hunk 5 above)
Comment (kdwyer) [reply addons/print/models/print_printer.py]: The existing code seems to assume integer inputs only, so I'm inclined to let responsibility for ensuring that remain with downstream code for now.  But you made me notice I should have catered for less than zero too, so thanks.

**Review by PeterC10 — APPROVED** [review]
LGTM

**Pull request merged**